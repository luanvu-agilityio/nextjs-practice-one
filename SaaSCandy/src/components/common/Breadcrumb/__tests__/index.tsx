import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import {
  Breadcrumb,
  BreadcrumbList,
  BreadcrumbItem,
  BreadcrumbLink,
  BreadcrumbPage,
  BreadcrumbSeparator,
  BreadcrumbEllipsis,
} from '../index';
import { forwardRef } from 'react';

describe('Breadcrumb Component', () => {
  describe('Snapshot Tests', () => {
    it('matches snapshot for light variant', () => {
      const { container } = render(
        <Breadcrumb variant='light'>
          <BreadcrumbList>
            <BreadcrumbItem variant='light'>
              <BreadcrumbLink href='/'>Home</BreadcrumbLink>
            </BreadcrumbItem>
            <BreadcrumbSeparator variant='light' />
            <BreadcrumbItem variant='light'>
              <BreadcrumbLink href='/products'>Products</BreadcrumbLink>
            </BreadcrumbItem>
            <BreadcrumbSeparator variant='light' />
            <BreadcrumbItem variant='light' active>
              <BreadcrumbPage>Current Page</BreadcrumbPage>
            </BreadcrumbItem>
          </BreadcrumbList>
        </Breadcrumb>
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot for dark variant', () => {
      const { container } = render(
        <Breadcrumb variant='dark'>
          <BreadcrumbList>
            <BreadcrumbItem variant='dark'>
              <BreadcrumbLink href='/'>Home</BreadcrumbLink>
            </BreadcrumbItem>
            <BreadcrumbSeparator variant='dark' />
            <BreadcrumbItem variant='dark' active>
              <BreadcrumbPage>Dashboard</BreadcrumbPage>
            </BreadcrumbItem>
          </BreadcrumbList>
        </Breadcrumb>
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot with ellipsis and custom separator', () => {
      const { container } = render(
        <Breadcrumb variant='light'>
          <BreadcrumbList>
            <BreadcrumbItem variant='light'>
              <BreadcrumbLink href='/'>Home</BreadcrumbLink>
            </BreadcrumbItem>
            <BreadcrumbSeparator variant='light'>
              <span>/</span>
            </BreadcrumbSeparator>
            <BreadcrumbItem variant='light'>
              <BreadcrumbEllipsis />
            </BreadcrumbItem>
            <BreadcrumbSeparator variant='light'>
              <span>/</span>
            </BreadcrumbSeparator>
            <BreadcrumbItem variant='light' active>
              <BreadcrumbPage>Current</BreadcrumbPage>
            </BreadcrumbItem>
          </BreadcrumbList>
        </Breadcrumb>
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('Interactive Tests', () => {
    it('renders all breadcrumb items and handles click interactions', async () => {
      const handleHomeClick = jest.fn((e: React.MouseEvent) =>
        e.preventDefault()
      );
      const handleProductsClick = jest.fn((e: React.MouseEvent) =>
        e.preventDefault()
      );
      const user = userEvent.setup();

      render(
        <Breadcrumb variant='light'>
          <BreadcrumbList>
            <BreadcrumbItem variant='light'>
              <BreadcrumbLink href='/' onClick={handleHomeClick}>
                Home
              </BreadcrumbLink>
            </BreadcrumbItem>
            <BreadcrumbSeparator variant='light' />
            <BreadcrumbItem variant='light'>
              <BreadcrumbLink href='/products' onClick={handleProductsClick}>
                Products
              </BreadcrumbLink>
            </BreadcrumbItem>
            <BreadcrumbSeparator variant='light' />
            <BreadcrumbItem variant='light' active>
              <BreadcrumbPage>Laptop</BreadcrumbPage>
            </BreadcrumbItem>
          </BreadcrumbList>
        </Breadcrumb>
      );

      // Check all items are rendered
      expect(screen.getByText('Home')).toBeInTheDocument();
      expect(screen.getByText('Products')).toBeInTheDocument();
      expect(screen.getByText('Laptop')).toBeInTheDocument();

      // Test click interactions
      await user.click(screen.getByText('Home'));
      expect(handleHomeClick).toHaveBeenCalledTimes(1);

      await user.click(screen.getByText('Products'));
      expect(handleProductsClick).toHaveBeenCalledTimes(1);

      // Current page should have aria-current="page"
      const currentPage = screen.getByText('Laptop');
      expect(currentPage).toHaveAttribute('aria-current', 'page');
      expect(currentPage).toHaveAttribute('aria-disabled', 'true');
    });

    it('renders custom separator and ellipsis correctly', () => {
      render(
        <Breadcrumb variant='light'>
          <BreadcrumbList>
            <BreadcrumbItem variant='light'>
              <BreadcrumbLink href='/'>Home</BreadcrumbLink>
            </BreadcrumbItem>
            <BreadcrumbSeparator variant='light'>
              <span data-testid='custom-separator'>/</span>
            </BreadcrumbSeparator>
            <BreadcrumbItem variant='light'>
              <BreadcrumbEllipsis />
            </BreadcrumbItem>
            <BreadcrumbSeparator variant='light' />
            <BreadcrumbItem variant='light' active>
              <BreadcrumbPage>Current</BreadcrumbPage>
            </BreadcrumbItem>
          </BreadcrumbList>
        </Breadcrumb>
      );

      // Custom separator should be rendered
      expect(screen.getByTestId('custom-separator')).toBeInTheDocument();

      // Ellipsis should be present with sr-only text
      expect(screen.getByText('More')).toBeInTheDocument();
      expect(screen.getByText('More')).toHaveClass('sr-only');
    });

    it('applies correct variant classes', () => {
      const { rerender } = render(
        <Breadcrumb variant='light' data-testid='breadcrumb'>
          <BreadcrumbList>
            <BreadcrumbItem variant='light'>
              <BreadcrumbLink href='/'>Home</BreadcrumbLink>
            </BreadcrumbItem>
          </BreadcrumbList>
        </Breadcrumb>
      );

      const breadcrumb = screen.getByTestId('breadcrumb');
      expect(breadcrumb).toHaveClass('text-slate-600');

      // Rerender with dark variant
      rerender(
        <Breadcrumb variant='dark' data-testid='breadcrumb'>
          <BreadcrumbList>
            <BreadcrumbItem variant='dark'>
              <BreadcrumbLink href='/'>Home</BreadcrumbLink>
            </BreadcrumbItem>
          </BreadcrumbList>
        </Breadcrumb>
      );

      expect(breadcrumb).toHaveClass('text-slate-300');
    });

    it('supports asChild prop with Slot component', () => {
      const CustomLink = forwardRef<
        HTMLAnchorElement,
        React.AnchorHTMLAttributes<HTMLAnchorElement>
      >((props, ref) => <a ref={ref} data-custom='true' {...props} />);
      CustomLink.displayName = 'CustomLink';

      render(
        <Breadcrumb variant='light'>
          <BreadcrumbList>
            <BreadcrumbItem variant='light'>
              <BreadcrumbLink asChild>
                <CustomLink href='/'>Home</CustomLink>
              </BreadcrumbLink>
            </BreadcrumbItem>
          </BreadcrumbList>
        </Breadcrumb>
      );

      const link = screen.getByText('Home');
      expect(link).toHaveAttribute('data-custom', 'true');
    });
  });
});
