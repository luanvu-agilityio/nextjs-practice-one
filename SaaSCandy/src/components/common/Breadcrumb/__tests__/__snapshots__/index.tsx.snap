// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Breadcrumb Component Snapshot Tests matches snapshot for dark variant 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="flex items-center gap-1.5 text-sm text-slate-300"
  >
    <ol
      class="flex flex-wrap items-center gap-1.5 break-words text-sm text-muted-foreground sm:gap-2.5"
    >
      <li
        class="inline-flex items-center gap-1.5 transition-colors hover:text-white"
      >
        <a
          class="transition-colors hover:text-foreground"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        aria-hidden="true"
        class="flex items-center text-slate-500"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5 transition-colors hover:text-white font-medium text-white"
      >
        <span
          aria-current="page"
          aria-disabled="true"
          class="font-normal text-foreground"
          role="link"
        >
          Dashboard
        </span>
      </li>
    </ol>
  </nav>
</div>
`;

exports[`Breadcrumb Component Snapshot Tests matches snapshot for light variant 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="flex items-center gap-1.5 text-sm text-slate-600"
  >
    <ol
      class="flex flex-wrap items-center gap-1.5 break-words text-sm text-muted-foreground sm:gap-2.5"
    >
      <li
        class="inline-flex items-center gap-1.5 transition-colors hover:text-slate-900"
      >
        <a
          class="transition-colors hover:text-foreground"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        aria-hidden="true"
        class="flex items-center text-slate-400"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5 transition-colors hover:text-slate-900"
      >
        <a
          class="transition-colors hover:text-foreground"
          href="/products"
        >
          Products
        </a>
      </li>
      <li
        aria-hidden="true"
        class="flex items-center text-slate-400"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5 transition-colors hover:text-slate-900 font-medium text-slate-900"
      >
        <span
          aria-current="page"
          aria-disabled="true"
          class="font-normal text-foreground"
          role="link"
        >
          Current Page
        </span>
      </li>
    </ol>
  </nav>
</div>
`;

exports[`Breadcrumb Component Snapshot Tests matches snapshot with ellipsis and custom separator 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="flex items-center gap-1.5 text-sm text-slate-600"
  >
    <ol
      class="flex flex-wrap items-center gap-1.5 break-words text-sm text-muted-foreground sm:gap-2.5"
    >
      <li
        class="inline-flex items-center gap-1.5 transition-colors hover:text-slate-900"
      >
        <a
          class="transition-colors hover:text-foreground"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        aria-hidden="true"
        class="flex items-center text-slate-400"
        role="presentation"
      >
        <span>
          /
        </span>
      </li>
      <li
        class="inline-flex items-center gap-1.5 transition-colors hover:text-slate-900"
      >
        <span
          aria-hidden="true"
          class="flex h-9 w-9 items-center justify-center"
          role="presentation"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-ellipsis h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="1"
            />
            <circle
              cx="19"
              cy="12"
              r="1"
            />
            <circle
              cx="5"
              cy="12"
              r="1"
            />
          </svg>
          <span
            class="sr-only"
          >
            More
          </span>
        </span>
      </li>
      <li
        aria-hidden="true"
        class="flex items-center text-slate-400"
        role="presentation"
      >
        <span>
          /
        </span>
      </li>
      <li
        class="inline-flex items-center gap-1.5 transition-colors hover:text-slate-900 font-medium text-slate-900"
      >
        <span
          aria-current="page"
          aria-disabled="true"
          class="font-normal text-foreground"
          role="link"
        >
          Current
        </span>
      </li>
    </ol>
  </nav>
</div>
`;
